### PR TITLE
Beautification+integration jack stuff

### PR DESCRIPTION
**Pull request by AntreasAntoniou**

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -91,124 +98,148 @@ def get_base_argument_parser():
...
+    data_loader,
+    model,
+    metric_tracker,
+    progress_panel,
+    progress_tracker,
+    overall_progress,
+    overall_task,
+):
+    model = model.train()
+    # with tqdm.tqdm(initial=0, total=len(data_loader), smoothing=0) as pbar:
```
Comment (jack-willturner) [inline train.py:112]: remove this tqdm comment

File: train.py [hunk 2]
```diff
@@ -329,66 +368,143 @@ def run_epoch(epoch, model, training, metric_tracker, data_loader):
...
+            border_style="green",
+            padding=(2, 2),
+        ),
+    )
+    progress_table.add_row(
+        Panel.fit(epoch_progress, title="[b]Jobs", border_style="red", padding=(1, 2))
+    )
+    progress_table.add_row(metric_tracker_train.per_epoch_table)
+    progress_table.add_row(metric_tracker_val.per_epoch_table)
+
```
Comment (jack-willturner) [inline train.py:417]: this whole section here is quite verbose and doesn't have much to do with training. can we move it into a utils function?

Comment (AntreasAntoniou) [reply train.py:417]: I was considering that too. I wanted to see what you thought before I added more abstraction.

Comment (AntreasAntoniou) [reply train.py:417]: I could easily have a whole class taking care of the whole interface thing. That would make everything a lot neater.

**Comment by jack-willturner** [comment]
I don't love having args outside of the main train file. Also excude_gpu_list is missing an l in exclude 

```
def add_extra_option_args(parser):

```
								
[utils/arg_parsing.py](https://github.com/bayeswatch/pytorch-experiments-template/blob/930d17f27753ecb510a34cf255e835ffac9c8694/utils/arg_parsing.py#L57) (Line 57)

File: utils/gpu_selection_utils.py [hunk 3]
```diff
@@ -28,4 +26,8 @@ def select_devices(
                 "the requirements or using num_gpus_to_use=0 to use CPU"
             )
 
-    return [str(gpu_idx) for gpu_idx in gpu_to_use]
+        os.environ["CUDA_VISIBLE_DEVICES"] = ",".join(
+            [str(gpu_idx) for gpu_idx in gpu_to_use]
+        )
+
+        print("GPUs selected have IDs {}".format(os.environ["CUDA_VISIBLE_DEVICES"]))
```
Comment (jack-willturner) [inline utils/gpu_selection_utils.py:33]: as always i would prefer an f-string here

File: utils/metric_tracking.py [hunk 4]
```diff
@@ -89,3 +120,64 @@ def save(self, overwrite=True):
...
+                        where_metrics = epochs == this_epoch
+                        v_mean = np.mean(v[where_metrics])
+                        v_std = np.std(v[where_metrics])
+                        epoch_metrics["{}_mean".format(k)].append(v_mean)
+                        epoch_metrics["{}_std".format(k)].append(v_std)
+        return epoch_metrics
+
+    def get_best_epoch_for_metric(self, metric_name, evaluation_metric=np.argmax):
+        return evaluation_metric(self.collect_per_epoch()[metric_name])
+
```
Comment (jack-willturner) [inline utils/metric_tracking.py:148]: I think having a single line function which just calls another function is a little dodgy. Better to just do this directly from wherever this is supposed to be called?

File: utils/storage.py [hunk 5]
```diff
@@ -9,8 +9,15 @@
...
 import json
 import os
-import glob
-import tarfile
+from rich import print
+
+
+def isfloat(x):
+    return isinstance(x, float)
+
```
Comment (jack-willturner) [inline utils/storage.py:17]: again don't like having single line functions. I also couldn't see anywhere in the code where this is called

File: notebooks/utils.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+def isint(x):
+    try:
+        a = float(x)
+        b = int(a) if a != float("Inf") else 9999999
+    except ValueError:
+        return False
+    else:
+        return a == b
+
```
Comment (jack-willturner) [inline notebooks/utils.py:21]: why not just import these from storage.py? or just do the comparison directly

**Review by jack-willturner — APPROVED** [review]

**Pull request merged**